return {
	all_units = {
		require("./confusion/core_state/Computed.spec"),
		require("./confusion/core_state/Value.spec"),
		require("./confusion/core_state/cast_to_state.spec"),
		require("./confusion/core_state/is_similar.spec"),
		require("./confusion/core_external/ExternalDebug.spec"),
		require("./confusion/core_external/ExternalLogger.spec"),
		require("./confusion/core_external/ExternalScheduler.spec"),
		require("./confusion/core_graph/change.spec"),
		require("./confusion/core_graph/depend.spec"),
		require("./confusion/core_graph/evaluate.spec"),
		require("./confusion/core_graph/is_eager.spec"),
		require("./confusion/core_graph/mark_dirty.spec"),
		require("./confusion/core_graph/observe_immediate.spec"),
		require("./confusion/core_graph/observe.spec"),
		require("./confusion/core_memory/cleanup.spec"),
		require("./confusion/core_memory/derive_scope.spec"),
		require("./confusion/core_memory/inner_scope.spec"),
		require("./confusion/core_memory/insert.spec"),
		require("./confusion/core_memory/scoped.spec"),
		require("./confusion/core_use/peek.spec"),
		require("./confusion/core_use/generic_use.spec"),

		require("./confusion/iter/Iter.spec"),
		require("./confusion/iter_for/ForKeys.spec"),
		require("./confusion/iter_for/ForValues.spec"),
		require("./confusion/iter_for/ForPairs.spec"),
		require("./confusion/iter_range/ForRange.spec"),

		require("./confusion/chrono/ExternalTime.spec"),
		require("./confusion/chrono/Timer.spec"),

		require("./confusion/lerp/lerp_type.spec"),
		require("./confusion/safe/Safe.spec"),
		require("./confusion/context/Contextual.spec"),
		require("./confusion/pipe/Pipe.spec"),
	},
}
